File: python/utils/deserialize_to_binary_tree.py
import sys

sys.path.append("../models")
from binary_tree_node import TreeNode
from typing import Optional
from collections import deque


# class TreeNode:
#     def __init__(self, val, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
#     def __repr__(self):
#         return 'TreeNode({})'.format(self.val)
def DeserializeFromList(vector: list[Optional[int]]) -> Optional[TreeNode]:
    """Deserialize a list to a binary tree by level order traversal (BFS)."""
    if len(vector) == 0:
        return None
    originalQueue: deque[Optional[TreeNode]] = deque()
    tempQueue: deque[Optional[TreeNode]] = deque()
    for i in range(len(vector)):
        if vector[i] == None:
            originalQueue.append(None)
        else:
            originalQueue.append(TreeNode(vector[i]))

    root: Optional[TreeNode] = originalQueue.popleft()
    tempQueue.append(root)

    while len(tempQueue) > 0:
        n: Optional[TreeNode] = tempQueue.popleft()
        if len(originalQueue) > 0:
            n.left = originalQueue.popleft()
            if n.left != None:
                tempQueue.append(n.left)
        if len(originalQueue) > 0:
            n.right = originalQueue.popleft()
            if n.right != None:
                tempQueue.append(n.right)

    return root


def deserializeFromString(string):
    if string == "{}":
        return None
    nodes = [
        None if val == "null" else TreeNode(int(val))
        for val in string.strip("[]{}").split(",")
    ]
    kids = nodes[::-1]
    root = kids.pop()
    for node in nodes:
        if node:
            if kids:
                node.left = kids.pop()
            if kids:
                node.right = kids.pop()
    return root


# # NOTE: need to install turtle
# def drawtree(root):
#     def height(root):
#         return 1 + max(height(root.left), height(root.right)) if root else -1
#
#     def jumpto(x, y):
#         t.penup()
#         t.goto(x, y)
#         t.pendown()
#
#     def draw(node, x, y, dx):
#         if node:
#             t.goto(x, y)
#             jumpto(x, y - 20)
#             t.write(node.val, align="center", font=("Arial", 12, "normal"))
#             draw(node.left, x - dx, y - 60, dx / 2)
#             jumpto(x, y - 20)
#             draw(node.right, x + dx, y - 60, dx / 2)
#
#     import turtle
#
#     t = turtle.Turtle()
#     t.speed(0)
#     turtle.delay(0)
#     h = height(root)
#     jumpto(0, 30 * h)
#     draw(root, 0, 30 * h, 40 * h)
#     t.hideturtle()
#     turtle.mainloop()


if __name__ == "__main__":
    arr: list[Optional[int]] = [1, 2, 2, 3, 3, None, None, 4, 4, None, None]
    root = DeserializeFromList(arr)
    if root == None:
        raise Exception("wrong")

    # deserializeFromString("[1,2,3,null,null,4,null,null,5]")
    # drawtree(deserializeFromString('[2,1,3,0,7,9,1,2,null,1,0,null,null,8,8,null,null,null,null,7]'))

File: rust/src/code/utils/deserialize_to_binary_tree.rs
use std::{cell::RefCell, collections::VecDeque, rc::Rc};

use crate::code::models::binary_tree_node::TreeNode;

#[allow(non_snake_case)]
/// Deserialize a vector to a binary tree by BFS
pub fn deserialize_to_BT(vector: Vec<Option<i32>>) -> Option<Rc<RefCell<TreeNode>>> {
    if vector.len() == 0 {
        return None;
    }

    let mut original_queue: VecDeque<Option<Rc<RefCell<TreeNode>>>> = VecDeque::new();
    let mut temp_queue: VecDeque<Option<Rc<RefCell<TreeNode>>>> = VecDeque::new();

    for i in 0..vector.len() {
        match vector[i] {
            Some(val) => {
                let node = Rc::new(RefCell::new(TreeNode::new(val)));
                original_queue.push_back(Some(node));
            }
            None => {
                original_queue.push_back(None);
            }
        }
    }

    if let Some(root) = original_queue.pop_front() {
        temp_queue.push_back(root.clone());
        while !temp_queue.is_empty() {
            if let Some(node) = temp_queue.pop_front() {
                if let Some(n) = node {
                    match original_queue.pop_front() {
                        Some(left) => {
                            n.borrow_mut().left = left.clone();
                            temp_queue.push_back(left.clone());
                        }
                        None => {
                            n.borrow_mut().left = None;
                        }
                    }
                    match original_queue.pop_front() {
                        Some(right) => {
                            n.borrow_mut().right = right.clone();
                            temp_queue.push_back(right.clone());
                        }
                        None => {
                            n.borrow_mut().right = None;
                        }
                    }
                }
            }
        }
        return root;
    }
    return None;
}
